app/src/combo.c

```diff
@@ -48,27 +66,66 @@ struct active_combo {
     // The keys are removed from this array when they are released.
     // Once this array is empty, the behavior is released.
     uint32_t key_positions_pressed_count;
-    struct zmk_position_state_changed_event
-        key_positions_pressed[CONFIG_ZMK_COMBO_MAX_KEYS_PER_COMBO];
+    struct zmk_position_state_changed_event key_positions_pressed[MAX_COMBO_KEYS];
 };
 
 struct combo_candidate {
-    const struct combo_cfg *combo;
+    uint32_t combo_idx;
     // the time after which this behavior should be removed from candidates.
     // by keeping track of when the candidate should be cleared there is no
     // possibility of accidental releases.
     int64_t timeout_at;
 };
 
+#define PROP_BIT_AT_IDX(n, prop, idx) BIT(DT_PROP_BY_IDX(n, prop, idx))
+
+#define NODE_PROP_BITMASK(n, prop)                                                                 \
+    COND_CODE_1(DT_NODE_HAS_PROP(n, prop),                                                         \
+                (DT_FOREACH_PROP_ELEM_SEP(n, prop, PROP_BIT_AT_IDX, (|))), (0))
+
+#define GET_KEY_POSITION_MASK_PORTION(idx, n) ((NODE_PROP_BITMASK(n, key_positions) >> idx) & 0xFF)
+
+#define COMBO_INST(n, positions)                                                                   \
+    COND_CODE_1(IS_EQ(DT_PROP_LEN(n, key_positions), positions),                                   \
+                (                                                                                  \
+                    {                                                                              \
+                        .timeout_ms = DT_PROP(n, timeout_ms),                                      \
+                        .require_prior_idle_ms = DT_PROP(n, require_prior_idle_ms),                \
+                        .key_positions = DT_PROP(n, key_positions),                                \
+                        .key_position_len = DT_PROP_LEN(n, key_positions),                         \
+                        .behavior = ZMK_KEYMAP_EXTRACT_BINDING(0, n),                              \
+                        .virtual_key_position = ZMK_VIRTUAL_KEY_POSITION_COMBO(__COUNTER__),       \
+                        .slow_release = DT_PROP(n, slow_release),                                  \
+                        .layer_mask = NODE_PROP_BITMASK(n, layers),                                \
+                    }, ),                                                                          \
+                ())
+
+#define COMBO_CONFIGS_WITH_MATCHING_POSITIONS_LEN(positions, _ignore)                              \
+    DT_INST_FOREACH_CHILD_VARGS(0, COMBO_INST, positions)
+static const struct combo_cfg combos[] = {
+    LISTIFY(10, COMBO_CONFIGS_WITH_MATCHING_POSITIONS_LEN, (), 0)};
```

Comment:
Should `10` be `MAX_COMBO_KEYS` instead? I have 30% confidence that I understand what this is doing, but it sounds like it is iterating over possible `key_positions` lengths.

Reply:
This deserved a comment: `MAX_COMBO_KEYS` is not a pre-processor constant, but a C union trick, so can't be used as a parameter to `LISTIFY`, so we basically pick an arbitrary high value that won't tax the pre-processor too much. I picked 10, since it's unlikely for someone to combo with more fingers than they have.

Reply:
I guess it doesn't hurt to set this to a high number, so maybe we go 'til 255 (devicetree limit, IIRC)? Unless you think preprocessor can slow down the build enough.

Reply:
> I guess it doesn't hurt to set this to a high number, so maybe we go 'til 255 (devicetree limit, IIRC)? Unless you think preprocessor can slow down the build by a significant amount.

It *might*, and honestly, this is iterating by key-positions *length*, so I'm having a hard time imagining we need to support a combo triggered by 256 different key positions?

Reply:
Right, just thought it was perhaps a number with lesser-magic involved than 10, it's fine to keep that too (perhaps with a comment and noting the limitation in docs).